Add support for userdata

jkmassel opened the pull request:
Adds support for passing a `$userdata` object into an `APNSRequest`, which the network layer will ensure is passed to the corresponding `APNSResponse` object. By default, the object will always contain the `apns_uuid` and `apns_token` fields, which are useful for associating the response with a given notification or device token. If instead of a UUID the developer is using an integer-based primary key, they can pass their own key in `$userdata` that allows for later retrieval.

**To Test:**
- The tests in this PR should be sufficient

jkmassel commented on src/model/APNSRequest.php:
```diff
@@ -48,6 +74,10 @@ public function get_url_for_configuration( APNSConfiguration $configuration ): s
 		return $configuration->get_endpoint() . $this->token;
 	}
 
+	public function get_userdata(): object {
```
Feel free to nitpick this name – I'm not in love with it, but I also couldn't come up with something better offhand 😅

mokagio commented on composer.json:
```diff
@@ -36,6 +36,7 @@
         "lint": "./vendor/bin/phpcs -s --ignore=tests/MockAPNSServer/node_modules/",
         "fix": "./vendor/bin/phpcbf --ignore=tests/MockAPNSServer/node_modules/",
         "test": "./vendor/bin/phpunit --coverage-clover 'coverage/coverage.xml' --coverage-html 'coverage'",
+        "quicktest": "./vendor/bin/phpunit --coverage-clover 'coverage/coverage.xml' --coverage-html 'coverage' --exclude e2e",
```
Nice

mokagio commented on tests/unit/APNSRequestTest.php:
```diff
@@ -18,6 +18,23 @@ public function testRequestInstantiationFromPayload() {
 		$this->assertEquals( $message, $this->decode( $request->get_body() )->aps->alert );
 	}
 
+	public function testThatUuidCannotBeOverwrittenWithCustomUserData() {
+		$valid_uuid = $this->random_uuid();
+		$metadata   = $this->new_metadata( 'topic', $valid_uuid );
+
+		$request = APNSRequest::from_string( '', '', $metadata, $this->new_userdata_with_uuid( $this->random_uuid() ) );
+
+		$this->assertEquals( $valid_uuid, $request->get_uuid() );
+	}
```
Nitpick. I'm not sure how random works in PHP (or anywhere else, really) but is there a chance that the two calls to `random_uuid()` here would result in the same value?

Would it be worth to do something like:

```swift
$valid_uuid = $this->random_uuid();
$different_uuid = $this->random_uuid();

$this->assertNotEqual( $valid_uuid, $different_uuid );
```

and then use `$different_uuid` to generate the `APNSRequest`?

The same observation applies to the next test.

jkmassel replied:
Interesting! I think this is rather unlikely – under the hood `random_uuid()` uses https://www.php.net/manual/en/function.random-bytes.php, which is said to be cryptographically secure, thus this shouldn't happen.

mokagio reviewed (approved):
Look good.

I'm curious to know why you chose to use an `object` instead of an `array` used as a dictionary for the user data.

When I think of a property where the user can pass custom values, using a dictionary is what comes to mind as the type for the job. But, I don't know PHP and I'm probably highly influenced by all the years working with Objective-C and Swift 😅

As for the name, I think user data is okay. It's sort of a standard name for this kind of functionality.

jkmassel commented:
>I'm curious to know why you chose to use an object instead of an array used as a dictionary for the user data.

This is a good question – originally I had it as an object because it used `serialize`, which meant that when the object comes out it would have the same type as when it went in. However, `serialize` can have security implications – while it _mostly_ doesn't matter here (the serialization isn't persisted anywhere, so object substitution attacks are basically impossible) I changed it to just marshal it to JSON and back instead, but I didn't update the types.

I agree with you, so I've changed it back to an array. I don't think it required another whole review, so I'll merge it at this point – thanks for bringing this up!

mokagio commented:
Thank you for the explanation 👍 .
